File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int m = matrix.size(), n = matrix[0].size();
        // if(!m || !n)    return false;
        int lo=0, hi=m*n-1, mid, i, j;
        while(lo<=hi){
            mid = (lo+hi)/2;
            i=mid/n, j=mid%n;
            if(target==matrix[i][j])
                return true;
            else if(target<matrix[i][j])
                hi = mid-1;
            else lo = mid+1;
        }
        return false;
    }
};

File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.java
class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        int m = matrix.length, n = matrix[0].length, lt = 0, rt = m * n - 1;
        while (lt <= rt) {
            int mid = (lt + rt) / 2;
            int i = mid / n, j = mid % n;
            if (target == matrix[i][j])
                return true;
            else if (matrix[i][j] < target)
                lt = mid + 1;
            else
                rt = mid - 1;
        }
        return false;
    }
}